Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -627,4 +627,132 @@ public struct NonBlockingFileIO: Sendable {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            try Posix.symlink(pathname: path, destination: destination)
+        }
+    }
+
+    /// Returns target of the symbolic link at `path` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the link to read.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` containing link target.
+    public func readlink(path: String, eventLoop: EventLoop) -> EventLoopFuture<String> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            let maxLength = Int(PATH_MAX)
+            let pointer = UnsafeMutableBufferPointer<CChar>.allocate(capacity: maxLength)
+            defer {
+                pointer.deallocate()
+            }
+            let length = try Posix.readlink(pathname: path, outPath: pointer.baseAddress!, outPathSize: maxLength)
+            return String(decoding: UnsafeRawBufferPointer(pointer).prefix(length), as: UTF8.self)
+        }
+    }
+
+    /// Removes symbolic link at `path` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the link to remove.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` which is fulfilled if the rename was successful or fails on error.
+    public func unlink(path: String, eventLoop: EventLoop) -> EventLoopFuture<Void> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            try Posix.unlink(pathname: path)
+        }
+    }
+
+    public struct DirectoryEntry {
+        // File number of entry
+        var ino: UInt64
+        // File type
+        var type: UInt8
+        // File name
+        var name: String
+    }
+
+    /// List contents of the directory at `path` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the directory to list the content of.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` containing the file list.
+    public func listDirectory(path: String, eventLoop: EventLoop) -> EventLoopFuture<[DirectoryEntry]> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            let dir = try Posix.opendir(pathname: path)
+            var entries: [DirectoryEntry] = []
+            do {
+                while let entry = try Posix.readdir(dir: dir) {
+                    let name = withUnsafeBytes(of: entry.pointee.d_name) { pointer -> String in
+                        // Per man readdir(3):
+                        // struct dirent {
+                        //  char d_name[256]; /* Null-terminated filename */
```

Comment:
Can you move this comment into the `os(Linux)` block and then an equivalent manual page comment from macOS into the other block?

Reply:
Relatedly @buttaface, do you know if Android behaves more like macOS or more like Linux here?

Reply:
It appears to [behave the same as linux](https://android.googlesource.com/platform/bionic/+/refs/heads/master/libc/include/dirent.h#71), so please add an Android conditional here too.

Reply:
should we just use `d_namlen` in all cases instead?

Reply:
actually, it seems that `d_name` is null-terminated on darwin too:
```
A directory entry has a struct dirent at the front of it, containing its
inode number, the length of the entry, and the length of the name
contained in the entry.  These are followed by the name padded to a 4
byte boundary with null bytes.  All names are guaranteed null terminated.
```

Reply:
converted to read null-terminated string in all cases